models/unittest/testdb.go

```diff
@@ -62,34 +63,45 @@ func InitSettings(extraConfigs ...string) {
 
 // TestOptions represents test options
 type TestOptions struct {
-	GiteaRootPath string
-	FixtureFiles  []string
-	SetUp         func() error // SetUp will be executed before all tests in this package
-	TearDown      func() error // TearDown will be executed after all tests in this package
+	FixtureFiles []string
+	SetUp        func() error // SetUp will be executed before all tests in this package
+	TearDown     func() error // TearDown will be executed after all tests in this package
 }
 
 // MainTest a reusable TestMain(..) function for unit tests that need to use a
 // test database. Creates the test database, and sets necessary settings.
-func MainTest(m *testing.M, testOpts *TestOptions) {
-	setting.CustomPath = filepath.Join(testOpts.GiteaRootPath, "custom")
-	InitSettings()
-
-	var err error
+func MainTest(m *testing.M, testOpts ...*TestOptions) {
+	_, file, _, _ := runtime.Caller(1)
+	var found bool
+	for {
+		file = filepath.Dir(file)
+		exist, _ := util.IsFile(filepath.Join(file, "go.mod")) // Gitea workspace should be only one golang project
+		if exist {
+			found = true
+			break
+		}
+	}
+	if !found {
+		panic("You cannot run the tests out of a golang repository")
+	}
```

Comment:
I do not think it needs `runtime.Caller`.

You could do the search from current working directory.

Reply:
The function maybe invoked from a deep sub package. What did you mean `current working directory`?

Reply:
IIRC Golang runs the tests in their directory (current working directory).

So the old code uses "../../.." to get the project directory, it's relative to current working directory. So it doesn't need to use `runtime.Caller` to get the source file's directory.